src/s390_ecc.c

```diff
@@ -1830,19 +1849,25 @@ struct {				\
 		if (k == NULL) {
 #ifdef ICA_FIPS
 			if (fips & ICA_FIPS_MODE) {
-				/* Random number must be invertible by the order of the curve.
-				 * Use one byte less */
 				fc |= 0x80; /* deterministic signature */
-				RAND_bytes(param.P521.rand + off + 1, sizeof(param.P521.rand) - off - 1);
+				do {
+					counter++;
+					if (RAND_bytes(param.P521.rand + off, sizeof(param.P521.rand) - off) != 1) {
+						rc = EIO;
+						break;
+					}
+					rc = s390_kdsa(fc, param.buff, NULL, 0);
```

Comment:
In order to avoid many retries caused by nonces that are greater than the curve order 'n' (cause CC=2 with KDSA) we need to do something special here for P-521:

For P-521 the order 'n' is `0x01fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffa51868783bf2f966b7fcc0148f709a5d03bb5c9b8899c47aebb6fb71e91386409`, thus, its first bytes is 0x01, so the first seven bits are off, only the 8th bit is on, resulting in a bit length of 521 bits (no wonder for P-521, right?).
Any random nonce that has anything else than 0x01 or 0x00 in the very first byte will be larger than the order, and thus KDSA will fail with CC=2 and the loop takes a retry. 

Add the following line just before the KDSA call to set the first bits to zero (maybe with a proper comment):
`*(param.P521.rand + off) &= 0x01;`
That way the random nonce is at least not larger than 521 bits. It could still be numerically larger than the order, but then the loop will run a retry, which is OK.

For the P-256 and P-384 cases this special logic is NOT require, because the order of those curves start with 0xffffff..., so all 8 bits of the first byte are used, and thus no making is required.